src/plugins/intel_gpu/src/plugin/program.cpp

```diff
@@ -488,6 +488,14 @@ void Program::InitProfileInfo(const std::string& layerName,
     perfEntry.parentPrimitive = parentId;
 }
 
+void Program::AddMemoryStateInfo(const std::string& variable_id, const cldnn::layout& layout) {
+    auto it = m_memoryStatesInfo.find(variable_id);
+    if (it != m_memoryStatesInfo.end())
+        it->second.insert(layout);
```

Comment:
Can there be multiple memories to same variable_id?

Reply:
`Program::AddMemoryStateInfo` aims to store information about variable layouts during primitives creation; hence neither stream nor network are created at this stage. This information about layouts is used later by graph to allocate variables memory in `InferRequest`. Just before inference the variables memories are assigned to the networks to be executed.
Unlike `cldnn::program`, `ov::runtime::intel_gpu::Program` is designed to handle dynamic batch and is accessible from cldnn primitive builders, therefore it's chosen to store information about variables layouts for all networks processing the batch.
According to the [documentation](https://docs.openvino.ai/2021.2/openvino_docs_IE_DG_network_state_intro.html#openvino_state_api), states should be associated with inference requests. Current [implementation](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_gpu/src/plugin/graph.cpp#L83) allows to reuse networks (and streams) to run different inference requests, therefore state memories cannot be assigned to networks because they may be corrupted by different inference requests. Let me demonstrate such behavior in the example:

```C++
// Unexpected behavior
auto& network = core.LoadNetwork(); // allocates memory for each network (stream)
...
inferRequest0.Infer(); // populates state memory
inferRequest1.Infer(); // overrides state memory populated by inferRequest0 since networks are shared between different inference requests
inferRequest0.Infer(); // state memory is corrupted by inferRequest1
```

With the existing solution each inference request has it's own copy of variables:
```C++
// Expected behavior
auto& network = core.LoadNetwork(); // stores information about variables layout
auto inferRequest0 = network.CreateInferRequest(); // allocates variables memories for inferRequest0
auto inferRequest1 = network.CreateInferRequest(); // allocates variables memories for inferRequest1
inferRequest0.Infer(); // populates state memory for inferRequest0
inferRequest1.Infer(); // populates state memory for inferRequest1
inferRequest0.Infer(); // reads memory populated by inferRequest0
```

Reply:
My point was not associating memory to network (The reason I mentioned network is to talk about our ongoing future feature dynamic shape, which is not available in current master branch).  It is okay with allocating memory per infer request, I agree that it should be done so. What I wanted to point out is that in the current GPU plugin, all streams should have same "layouts". Even with dynamic batch, all the streams should have same layout for now. So I wondered why we should have such a multiple layouts for one variable.

Reply:
In the current implementation of dynamic batching the shapes of inputs and states are different for different networks processing the same batch. This allows to execute the whole batch in the most efficient way. The Graph component creates a network for each power of two batch size keeping other dimensions the same and executes only those of them which are required to process input with shape size known only during execution. Let's take a look at example:

Let's have an IR model with input of shape (1..16, 3, 16, 16) where batch size can take dynamic values 1..16. Networks don't support dynamic batches, therefore for such a model Graph creates 5 networks with the following input shapes:
(1,   3, 16, 16)
(2,   3, 16, 16)
(4,   3, 16, 16)
(8,   3, 16, 16)
(16, 3, 16, 16)

During execution inference request "knows" shape of the input. Let's suppose it's (7, 3, 16, 16). In this case inference request will run only the following networks and concatenate results by batch axis:
(1, 3, 16, 16)
(2, 3, 16, 16)
(4, 3, 16, 16)

On the next inference input can be (8, 3, 16, 16) and only one network will be invoked:
(8, 3, 16, 16)

The same logic is applicable to variables states and this is the reason why there are multiple layouts for the same variable.

Reply:
I was misunderstanding how dynamic shapes are being dealt with, so far I thought the streams are unified. Thanks for the detailed information! I think once the full dynamic shape is achieved current dynamic batch is no longer needed, but for now, it make sense. Thanks for the effort : )